### PR TITLE
core: Rename ```typ``` properties to ```type```

### PR DESCRIPTION
**Pull request by kingiler**
Address #1174

### CODE REVIEW

File: docs/Toy/toy/dialects/toy.py [hunk 1]
```diff
@@ -3,58 +3,47 @@
...
-    Block,
-    Region,
-    OpResult,
-)
-
 from xdsl.dialects.builtin import (
+    DenseIntOrFPElementsAttr,
     Float64Type,
     FunctionType,
+    StringAttr,
```
Comment (compor) [inline docs/Toy/toy/dialects/toy.py:14]: Can you please adjust your IDE/linters/etc to avoid reordering imports?
It adds a lot of noise that is unrelated to the changes in this PR

**Comment by AntonLydike** [comment]
I'm curious why this PR also re-orders all imports, are you running a custom formatting script?

It would help immensely with reviewing if we only had relevant changes in the diff!

**Comment by math-fehr** [comment]
I'll look at it once the imports changes are removed, but otherwise it looks good to me!

**Comment by kingiler** [comment]
The import changes issue is addressed in #1243. Now the only changes should the the apis.

File: docs/Toy/toy/dialects/toy.py (hunk 1 above)
Comment (kingiler) [reply docs/Toy/toy/dialects/toy.py]: Addressed in #1243. Rebased to main.

File: tests/dialects/test_mpi_lowering.py [hunk 2]
```diff
@@ -32,8 +32,8 @@ def check_emitted_function_signature(
         # check that the argument type is correct (if constraint present)
         if typ is not None:
```
Comment (adutilleul) [inline tests/dialects/test_mpi_lowering.py:33]: This should probably be renamed `type` also.

File: xdsl/dialects/arith.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def __init__(self, value: AnyIntegerAttr | FloatAttr[AnyFloat]):
         if isinstance(value, FloatAttr):
             typ = value.type
         else:
-            typ = value.typ
+            typ = value.type
```
Comment (adutilleul) [inline xdsl/dialects/arith.py:65]: The variable should also be named `type` here.

File: xdsl/dialects/func.py [hunk 4]
```diff
@@ -378,7 +380,7 @@ def parse(cls: type[Return], parser: Parser) -> Return:
                 parser.raise_error("Expected the same number of types and arguments!")
             for arg, typ in zip(args, types):
```
Comment (adutilleul) [inline xdsl/dialects/func.py:381]: Also `type` here.

File: xdsl/dialects/memref.py [hunk 5]
```diff
@@ -177,21 +177,21 @@ class Load(IRDLOperation):
...
 
-        if self.memref.typ.get_num_dims() != len(self.indices):
+        if self.memref.type.get_num_dims() != len(self.indices):
             raise Exception("expected an index for each dimension")
 
     @staticmethod
     def get(ref: SSAValue | Operation, indices: Sequence[SSAValue | Operation]) -> Load:
         ssa_value = SSAValue.get(ref)
-        typ = ssa_value.typ
+        typ = ssa_value.type
```
Comment (adutilleul) [inline xdsl/dialects/memref.py:194]: Another `type` issue here.

File: xdsl/dialects/printf.py [hunk 6]
```diff
@@ -68,7 +68,7 @@ def parse(cls: type[PrintFormatOp], parser: Parser) -> PrintFormatOp:
             args.append(arg := parser.parse_operand())
             parser.parse_characters(":", " - all arguments must have a type")
             typ = parser.parse_type()
```
Comment (adutilleul) [inline xdsl/dialects/printf.py:70]: Yet another `type` issue here.

File: xdsl/dialects/scf.py [hunk 7]
```diff
@@ -233,7 +233,7 @@ def verify_(self) -> None:
         # operand type)
         for reduction in range(num_reductions):
             typ = self.get_arg_type_of_nth_reduction_op(reduction)
```
Comment (adutilleul) [inline xdsl/dialects/scf.py:235]: Yet another `type` issue here. (x2)

File: xdsl/dialects/scf.py [hunk 8]
```diff
@@ -198,7 +198,7 @@ def verify_(self) -> None:
         num_reductions = self.count_number_reduction_ops()
         for reduction in range(num_reductions):
             typ = self.get_arg_type_of_nth_reduction_op(reduction)
```
Comment (adutilleul) [inline xdsl/dialects/scf.py:200]: Yet another `type` issue here. (x1)

File: xdsl/printer.py [hunk 9]
```diff
@@ -675,20 +675,22 @@ def print_op_with_default_format(self, op: Operation) -> None:
...
         # Print the operation type
         self.print(" : (")
-        self.print_list(op.operands, lambda operand: self.print_attribute(operand.typ))
+        self.print_list(op.operands, lambda operand: self.print_attribute(operand.type))
         self.print(") -> ")
         if len(op.results) == 0:
             self.print("()")
         elif len(op.results) == 1:
-            typ = op.results[0].typ
+            typ = op.results[0].type
```
Comment (adutilleul) [inline xdsl/printer.py:683]: Yet again another `type` issue.

File: tests/dialects/test_llvm.py [hunk 10]
```diff
@@ -86,7 +86,7 @@ def test_llvm_getelementptr_op_invalid_construction():
 def test_llvm_getelementptr_op():
     size = arith.Constant.from_int_and_width(1, 32)
     ptr = llvm.AllocaOp.get(size, builtin.i32)
-    ptr_typ = llvm.LLVMPointerType.typed(ptr.res.typ)
+    ptr_typ = llvm.LLVMPointerType.typed(ptr.res.type)
```
Comment (adutilleul) [inline tests/dialects/test_llvm.py:89]: should be named `ptr_type` here

File: xdsl/dialects/gpu.py [hunk 11]
```diff
@@ -150,8 +150,8 @@ class AllocOp(IRDLOperation):
 
     def verify_(self) -> None:
         ndyn = len(self.dynamicSizes)
-        assert isinstance(self.result.typ, memref.MemRefType)
-        typ: memref.MemRefType[Attribute] = self.result.typ
+        assert isinstance(self.result.type, memref.MemRefType)
+        typ: memref.MemRefType[Attribute] = self.result.type
         ndyn_typ = len([i for i in typ.shape.data if i.value.data == -1])
```
Comment (adutilleul) [inline xdsl/dialects/gpu.py:155]: should be named `ndyn_type` here

File: docs/Toy/toy/dialects/toy.py [hunk 12]
```diff
@@ -230,7 +230,7 @@ def verify_(self):
             raise VerifyException("Expected last op of FuncOp to be a ReturnOp")
 
         operand = last_op.input
-        operand_typ = None if operand is None else operand.typ
+        operand_typ = None if operand is None else operand.type
 
         return_typs = self.function_type.outputs.data
```
Comment (adutilleul) [inline docs/Toy/toy/dialects/toy.py:235]: should be named `return_types` here

**Comment by webmiche** [comment]
Let me repeat my recent comment on the issue here: 
> Why do we want to do this? `type` is a keyword right? Feels better to stay with `typ`, IMO :thinking:

**Review by adutilleul — CHANGES_REQUESTED** [review]
If we want to be consistent, you should also rename all the stray variable names containing `typ` or `typs`, not sure I catched all of them in my review but you should take a look at this.

**Comment by adutilleul** [comment]
> Let me repeat my recent comment on the issue here:
> 
> > Why do we want to do this? `type` is a keyword right? Feels better to stay with `typ`, IMO thinking

`type` is a built-in function, so you could prefix `__builtin__`  (__builtin__.type) to still use it but that might be confusing indeed

**Comment by math-fehr** [comment]
> Let me repeat my recent comment on the issue here:
> 
> > Why do we want to do this? `type` is a keyword right? Feels better to stay with `typ`, IMO 🤔

`type` is not a keyword, right? It's a builtin function though. I think it should be safe to have `type` as a field, as it anyway needs `self.type` to access it.

I personally think that we should use `type` rather than `typ` for fields, which is the case here. What do you think?

**Comment by webmiche** [comment]
> > Let me repeat my recent comment on the issue here:
> > > Why do we want to do this? `type` is a keyword right? Feels better to stay with `typ`, IMO thinking
> 
> `type` is not a keyword, right? It's a builtin function though. I think it should be safe to have `type` as a field, as it anyway needs `self.type` to access it.
> 
> I personally think that we should use `type` rather than `typ` for fields, which is the case here. What do you think?

Okay, sure. I think my point still stands though. I feel that we will run into cases where this will get annoying, so I would try to circumvent it and do `typ` from the get go. But I don't have a strong opinion on it. If you feel `type` is the way, I am happy to take `type`.

**Comment by kingiler** [comment]
I am thinking about the scope of this PR. I think having a class field named ```type``` is fine. But for the others variables named ```typ```, should they put into next PR?

**Comment by math-fehr** [comment]
> I am thinking about the scope of this PR. I think having a class field named `type` is fine. But for the others variables named `typ`, should they put into next PR?

I think that's exacly @michel-steuwer point here. I believe class fields should be named `type`, as they don't shadow anything. However, variables named `typ` might be a bit dangerous, and that's where I believe Michel wants to avoid them. I'm fine with both for the variables, as anyway we have pyright telling us that we are doing something wrong.

**Review by superlopuh — APPROVED** [review]
I closed the issue before seeing this PR, and I actually think that the diff is a strict improvement. I don't feel very strongly about this, but I think we should merge this PR.

**Review by adutilleul — APPROVED** [review]
